raft/src/main/java/org/apache/kafka/raft/LeaderState.java

```diff
@@ -79,6 +86,39 @@ protected LeaderState(
         this.grantingVoters = Collections.unmodifiableSet(new HashSet<>(grantingVoters));
         this.log = logContext.logger(LeaderState.class);
         this.accumulator = Objects.requireNonNull(accumulator, "accumulator must be non-null");
+        // use the 1.5x fetch timeout to tolerate some network transition time or other IO time.
+        this.fetchTimeoutMs = (int) (fetchTimeoutMs * 1.5);
+        this.fetchTimer = time.timer(fetchTimeoutMs);
+    }
+
+    // Check if the fetchTimer is expired because we didn't receive a valid fetch/fetchSnapshot request from the majority of
+    // the voters within fetch timeout.
+    public boolean hasMajorityFollowerFetchExpired(long currentTimeMs) {
+        fetchTimer.update(currentTimeMs);
+        boolean isExpired = fetchTimer.isExpired();
+        if (isExpired) {
+            log.info("Did not receive fetch request from the majority of the voters within {}ms. Current fetched voters are {}.",
+                    fetchTimeoutMs, fetchedVoters);
+        }
+        return isExpired;
+    }
+
+    // Reset the fetch timer if we've received fetch/fetchSnapshot request from the majority of the voter
+    public void maybeResetMajorityFollowerFetchTimer(int id, long currentTimeMs) {
+        updateFetchedVoters(id);
+        // The majority number of the voters excluding the leader. Ex: 3 voters, the value will be 1
+        int majority = voterStates.size() / 2;
+        if (fetchedVoters.size() >= majority) {
+            fetchedVoters.clear();
+            fetchTimer.update(currentTimeMs);
+            fetchTimer.reset(fetchTimeoutMs);
+        }
+    }
+
+    private void updateFetchedVoters(int id) {
+        if (isVoter(id)) {
```

Comment:
We should be defensive against this getting called with the local replica. Let's throw an `IllegalArgumentException` if `id` is equal to the `localId`.